xformers/components/multi_head_dispatch.py

```diff
@@ -298,11 +306,10 @@ def check(t, name):
         y = self.attention(q, k, v, **kw_mask_args)
 
         # Re-assemble all head outputs side by side
-        y = (
-            y.view(B, self.num_heads, S_Q, self.dim_value_head)
-            .transpose(1, 2)
-            .flatten(start_dim=2, end_dim=3)
-        )
+        y = y.view(B, self.num_heads, S_Q, self.dim_value_head)
+        if not self.attention.supports_b_s_h_d:
+            y = y.transpose(1, 2)
+        y = y.flatten(start_dim=2, end_dim=3)
```

Comment:
```suggestion
        if not self.attention.supports_b_s_h_d:
            y = (
                y.view(B, self.num_heads, S_Q, self.dim_value_head)
                .transpose(1, 2)
            )
        y = y.flatten(start_dim=2, end_dim=3)
```

Of course, we need to make the assumption that all attentions that accept [B, S, H, D] as input return [B, S, H, D] too (seems true for both of ours). When `self.attention.supports_b_s_h_d` is True, the output should be already in correct [B, S, H, D] form. We could add the view to make that super explicit. However, current `y.view(B, self.num_heads, S_Q, self.dim_value_head)` would break the result.

Reply:
This is the explicit suggestion which should basically be a no-op (at least from CUDA pov), but makes clear what we're flattening afterwards.

```suggestion
        if self.attention.supports_b_s_h_d:
           y = y.view(B, S_Q, self.num_heads, self.dim_value_head)
        else:
            y = (
                y.view(B, self.num_heads, S_Q, self.dim_value_head)
                .transpose(1, 2)
            )
        y = y.flatten(start_dim=2, end_dim=3)
```